### PR TITLE
hugolib: Include path prefix to generated links

### LINKED ISSUE #2227: Relative links to non-page files do not include path prefix
When using the relative files feature (setting the blackfriday `sourceRelativeLinksEval` option to true), the generated links do not have the path prefix from baseURL included. This means the links break if the site root is not the webserver root.

As an example, say that there are three files in content:

``` content/pages/page1.md
content/pages/page2.md
content/files/image1.png
```

and page1.md contains the text

```
[link to page 2](page2.md)
![](/files/image1.png)
```

and the config has `baseurl = "http://mysite.com/myroot/"`.

Then the link to page2.md will be `"/myroot/pages/page2/"`, while the image src will be "/files/image1.png" instead of the correct `"/myroot/files/image1.png"`.

I figure the `SourceRelativeLinkFile` function needs to learn about prefixes, but I don't know enough about the Hugo data structures to teach it...

This is with Hugo 0.16.

### PR DESCRIPTION
**Pull request by ks888**
Fixes #2227

### CODE REVIEW

**Comment by SvenDowideit** [comment]
#2227 isn't a bug. we use `sourceRelativeLinksEval` to allow us to relocate the output html to different directory locations. 

This also means that we only use a leading `/` when we want to break out of the relative linking.

**Comment by anthonyfok** [comment]
I think I can reproduce the problem described in #2227.  @SvenDowideit, please take a second look (like actually test it?) before dismissing it as a non-bug.  Thanks!

**Comment by SvenDowideit** [comment]
its not quite that simple - the original report in #2227 isn't a bug - its how `sourceRelativeLinksEval=true` should work.

there are subsequent comments in that issue that show a horrible cascade of related issues - which I am trying to work out

but... I'm almost certain that adding the baseUrl path in `SourceRelativeLinkFile()` isn't right - but for that, we need a heck of alot more tests than we have.

**Comment by SvenDowideit** [comment]
https://github.com/spf13/hugo/issues/2227#issuecomment-231013238 seems apply to this PR :/

File: hugolib/site.go [hunk 1]
```diff
@@ -343,8 +343,19 @@ func (s *SiteInfo) SourceRelativeLinkFile(ref string, currentPage *Page) (string
 			return "", fmt.Errorf("No file found for \"%s\" on page \"%s\".\n", ref, currentPage.Source.Path())
 		}
 
-		link = target.Path()
-		return "/" + filepath.ToSlash(link), nil
+		link = filepath.ToSlash(target.Path())
+		if viper.GetBool("CanonifyURLs") {
+			return "/" + link, nil
```
Comment (SvenDowideit) [inline hugolib/site.go]: @bep / @anthonyfok from what I have gleaned from the docs, `CanonifyURLs` should mean that the baseUrl _is_ added?

https://gohugo.io/extras/urls#canonicalization

interestingly, when I fix my SourceRelativeLink() to _not_  unintentionally add the `baseUrl` path to pages, `CanonifyURLs=true` starts fixing up abs urls by adding them. (yeah, convoluted)

**Comment by SvenDowideit** [comment]
So I think #2277 fixes this properly for most combinations of `canonifyURLs`, `relativeURLs` and `sourceRelativeLinksEval` and `baseurl`

but because the baseUrl part is added (er, i think) in a transform at the end, I have no idea how to write a test for it)

**Comment by ks888** [comment]
Yes, I agree with @SvenDowideit
> but... I was almost certain that adding the baseUrl path in SourceRelativeLinkFile() isn't right - but for that, we need a heck of alot more tests than we have.

and it's ok for me to close this PR.

**Pull request closed without merging**